### PR TITLE
Expose JsonSerializer.IsReadOnly and MakeReadOnly() APIs.

### LINKED ISSUE #54482: Provide a public api to "lock" a `JsonSerializerOptions`
## Background and Motivation

We have a set of contract libraries that we create, all of which depend on a "core" contract library, which defines the json serialization options as a static property. Currently, this property is internal, and we use `InternalsVisibleTo` to give access to this property to the actual contracts projects, but it would be preferable to just have this be a public property. 

The reason we don't just have it as public is that we don't want consumers of our contracts to modify it (as it's a shared static instance). However, reading through the source of `JsonSerializerOptions` I noticed that it contains a [VerifyMutable method](https://github.com/dotnet/runtime/blob/f891033db5b8ebf651176a3dcc3bec74a217f85e/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L676), meaning that there are cases where a JsonSerializerOptions is *not* mutable. I suggest enabling a simple public API that can take a given `JsonSerializerOptions` and "lock" it, or "make it immutable", so that it's safe to share in library code. This would be a no-op if the options is already immutable.

I do not know what name would be best for such a method, so the one below is just a suggestion.

## Proposed API

```diff
namespace System.Text.Json
{
    public partial class JsonSerializerOptions 
    {
+       public bool IsImmutable { get; }
+       public void EnsureImmutable(); 
    }
```

## Usage Examples

```csharp
var options = new JsonSerializerOptions { TypeInfoResolver = new MyFancyResolver(), WriteIndented = true };

Console.WriteLine(options.IsLocked); // False
var converter = options.GetConverter(typeof(MyPoco)); // Returns an uncached converter instance

options.Lock(); // lock the instance and initialize the metadata cache

Console.WriteLine(options.IsLocked); // True
converter = options.GetConverter(typeof(MyPoco)); // Returns a cached converter instance
```

## Alternative Designs

We might want to consider exposing equivalent functionality to JsonTypeInfo and JsonPropertyInfo, whose design uses a similar locking design. We're also not certain about naming, the obvious alternative being `IsLocked`/`Lock()` but there are concerns that this connotes with exclusive locking.

## Risks

I don't see any risk with this API. It would also help codify the fact that a JsonSerializerOptions can be "locked" (today it just happens implicitly).

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #54482.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs [hunk 1]
```diff
@@ -24,13 +24,25 @@ public abstract partial class JsonSerializerContext : IJsonTypeInfoResolver
...
+                JsonSerializerOptions? options = _options;
+
+                if (options is null)
+                {
+                    options = new JsonSerializerOptions { TypeInfoResolver = this };
+                    options.MakeReadOnly();
+                    _options = options;
+                }
+
+                return options;
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:38]: I'm wondering if you should instead do something like:

```suggestion
                    _options ??= options;
                }

                return _options;
```

to ensure we return the same instance in the multi-threaded scenarios

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs [hunk 2]
```diff
@@ -602,20 +602,35 @@ public bool WriteIndented
...
+        /// has been associated with a <see cref="JsonSerializerContext"/> instance,
+        /// or a user explicitly called the <see cref="MakeReadOnly"/> method on the instance.
+        /// </remarks>
+        public bool IsReadOnly => _isReadOnly;
+
+        /// <summary>
+        /// Locks the current instance for further modification.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The instance does not specify a <see cref="TypeInfoResolver"/> setting.</exception>
+        /// <remarks>This method is idempotent.</remarks>
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:625]: can we use simple english here? i.e. `This method can be safely called multiple times.`

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs [hunk 3]
```diff
@@ -602,20 +602,35 @@ public bool WriteIndented
...
+        public void MakeReadOnly()
         {
-            get => _isImmutable;
-            set
+            if (_typeInfoResolver is null)
             {
-                Debug.Assert(value, "cannot unlock options instances");
-                Debug.Assert(_typeInfoResolver != null, "cannot lock without a resolver.");
-                _isImmutable = true;
+                ThrowHelper.ThrowInvalidOperationException_JsonSerializerOptionsNoTypeInfoResolverSpecified();
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:630]: should this be assigning default resolver instead? I guess we can re-iterate on that in the future if we get feedback but I'd prefer more intuitive behavior than linker safety in this case

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/OptionsTests.cs [hunk 4]
```diff
@@ -135,11 +135,47 @@ public static void NewDefaultOptions_TypeInfoResolverIsNull()
...
 
+        [Fact]
+        public static void NewDefaultOptions_IsNotReadOnly()
+        {
+            var options = new JsonSerializerOptions { TypeInfoResolver = new DefaultJsonTypeInfoResolver() };
+            Assert.False(options.IsReadOnly);
+
+            options.MakeReadOnly();
+            Assert.True(options.IsReadOnly);
+            options.MakeReadOnly(); // Is idempotent operation
```
Comment (krwq) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/OptionsTests.cs:146]: you might wanna double and triple check IsReadOnly

**Review by krwq — APPROVED** [review]
LGTM with some small comments - I'll only ask to change the docs to not use word `indempotent` remaining comments fix as you see fit

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:630]: No -- we would need to mark the method as linker-unsafe, and I'd rather avoid doing that in cases other than ones mandated by backward compatibility.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:625]: The term idempotency is fairly common and well-understood. This includes terminology in our own docs, see:

https://grep.app/search?q=idempoten&filter[repo.pattern][0]=dotnet/docs

or

https://github.com/dotnet/dotnet-api-docs/search?q=idempotent

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:38]: We can't really guarantee that the same instance gets returned in multi-threaded scenarios unless we use interlocked operations, but we don't really need to do that either.

**Pull request merged**